libopae/src/error.c

```diff
@@ -0,0 +1,286 @@
+
+	struct error_list *p = _token->errors;
+	while (p) {
+		if (i == error_num) {
+			// test if file exists
+			if (stat(p->error_file, &st) == -1) {
+				FPGA_MSG("can't stat %s", p->error_file);
+				return FPGA_EXCEPTION;
+			}
+			res = sysfs_read_u64(p->error_file, value);
+			if (res != FPGA_OK) {
+				FPGA_MSG("can't read error file '%s'", p->error_file);
+				return res;
+			}
+
+			return FPGA_OK;
+		}
+		i++;
+		p = p->next;
+	}
+
+	FPGA_MSG("error %d not found", error_num);
+	return FPGA_NOT_FOUND;
+}
+
+fpga_result fpgaClearError(fpga_token token, uint32_t error_num)
+{
+	struct _fpga_token *_token = (struct _fpga_token *)token;
+	struct stat st;
+	uint32_t i = 0;
+	uint64_t value = 0;
+	fpga_result res = FPGA_OK;
+
+	// TODO: make thread-safe
+
+	struct error_list *p = _token->errors;
+	while (p) {
+		if (i == error_num) {
+			if (!p->info.can_clear) {
+				FPGA_MSG("can't clear error '%s'", p->info.name);
+				return FPGA_NOT_SUPPORTED;
+			}
+			// read current error value
+			res = fpgaReadError(token, error_num, &value);
+			if (res != FPGA_OK)
+				return res;
+
+			// write to 'clear' file
+			if (stat(p->clear_file, &st) == -1) {
+				FPGA_MSG("can't stat %s", p->clear_file);
+				return FPGA_EXCEPTION;
+			}
+			res = sysfs_write_u64(p->clear_file, value);
+			if (res != FPGA_OK) {
+				FPGA_MSG("can't write clear file '%s'", p->clear_file);
+				return res;
+			}
+			return FPGA_OK;
+		}
+		i++;
+		p = p->next;
+	}
+
+	FPGA_MSG("error info %d not found", error_num);
+	return FPGA_NOT_FOUND;
+}
+
+fpga_result fpgaClearAllErrors(fpga_token token)
+{
+	struct _fpga_token *_token = (struct _fpga_token *)token;
+	uint32_t i = 0;
+	fpga_result res = FPGA_OK;
+
+	// TODO: make thread-safe
+
+	struct error_list *p = _token->errors;
+	while (p) {
+		// if error can be cleared
+		if (p->info.can_clear) {
+			// clear error
+			res = fpgaClearError(token, i);
+			if (res != FPGA_OK)
+				return res;
+		}
+		i++;
+		p = p->next;
+	}
+
+	return FPGA_OK;
+}
+
+fpga_result fpgaGetErrorInfo(fpga_token token,
+			     uint32_t error_num,
+			     struct fpga_error_info *error_info)
+{
+	struct _fpga_token *_token = (struct _fpga_token *)token;
+	uint32_t i = 0;
+
+	if (!error_info) {
+		FPGA_MSG("error_info is NULL");
+		return FPGA_INVALID_PARAM;
+	}
+
+	// TODO: make thread-safe
+
+	// TODO: should we populate this if user hasn't created a
+	//       properties structure for this token?
+
+	struct error_list *p = _token->errors;
+	while (p) {
+		if (i == error_num) {
+			memcpy_s(error_info, sizeof(struct fpga_error_info), &p->info, sizeof(struct fpga_error_info));
+			return FPGA_OK;
+		}
+		i++;
+		p = p->next;
+	}
+
+	FPGA_MSG("error info %d not found", error_num);
+	return FPGA_NOT_FOUND;
+}
+
+/* files and directories to ignore when looking for errors */
+const char *errors_exclude[] = {
+	"revision",
+	"uevent",
+	"power",
+	"clear"
+};
+#define NUM_ERRORS_EXCLUDE (sizeof(errors_exclude) / sizeof(char *))
```

Comment:
Wouldn't it be better to more explicit on what nodes we operate on and whitelist the files? 
This would make it easier to determine how to clear errors too (without relying on the existence of a "clear" file).

Reply:
I was looking for something that minimizes maintenance if people start adding additional errors. Perhaps being more explicit about the naming of clearable errors would help (like "errors" and "errors_clear")?